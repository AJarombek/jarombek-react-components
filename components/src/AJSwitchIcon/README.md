### Overview

Wrapper around the switch component which has two icons which label the two different switch states.

**Use cases:**

```javascript
import { AJSwitchIcon } from 'jarombek-react-components';

const CustomComponent = () => 
  <AJSwitchIcon
    offImageUrl="https://asset.jarombek.com/off.png"
    onImageUrl="https://asset.jarombek.com/on.png"
    onChange={f=>f}
    initialState={true}
    disabled={false}
  />
```

### Files

| Filename                   | Description                                                                     |
|----------------------------|---------------------------------------------------------------------------------|
| `AJSwitchIcon.js`          | The reusable `AJSwitchIcon` component.                                          |
| `index.js`                 | Index file which exports the `AJSwitchIcon` component.                          |
| `styles.js`                | JSS styles for the `AJSwitchIcon` component.                                    |
