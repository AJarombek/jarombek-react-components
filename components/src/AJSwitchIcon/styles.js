/**
 * Stylesheet for the AJSwitchIcon component using JSS.
 * Author: Andrew Jarombek
 * Date: 1/1/2020
 */

export default {
  ajSwitchIcon: {

  },
  ajSwitchOffImage: {

  },
  ajSwitchOnImage: {

  }
};
