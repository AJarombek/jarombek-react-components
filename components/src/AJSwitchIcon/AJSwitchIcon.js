/**
 * Wrapper around the reusable switch component that displays an image for each state.
 * @author Andrew Jarombek
 * @since 1/1/2020
 */

import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {createUseStyles} from 'react-jss';
import classnames from 'classnames';

import styles from './styles';
import AJSwitch from '../AJSwitch/AJSwitch';

const useStyles = createUseStyles(styles);

/**
 * Component representing a toggleable switch with icons for each state.
 * @param offImageUrl The image representing the 'off' state of the switch.
 * @param onImageUrl The image representing the 'on' state of the switch.
 * @param onChange Function that is called when the switch is toggled.  The function takes a single
 * boolean argument, whose value is {@code true} if the switch is on and {@code false} if the
 * switch is off.
 * @param initialState The initial on/off state of the switch.
 * @param disabled Whether clicking on the switch changes its state.
 * @return {*} React elements representing a toggleable switch with icons.
 */
const AJSwitchIcon = ({offImageUrl, onImageUrl, onChange, initialState=false, disabled=false}) => {
  const classes = useStyles();

  /**
   * Generate the 'off' icon classes based on whether the switch is on or off.
   * @param state The state of the AJSwitch child component.
   * @return {*} The classes attached to the 'off' icon.
   */
  const getOffIconClasses = (state) => {
    return state ?
      classnames(classes.ajSwitchIcon, classes.ajSwitchIconNotSelected, 'aj-switch-icon-off') :
      classnames(classes.ajSwitchIcon, 'aj-switch-icon-off');
  };

  /**
   * Generate the 'on' icon classes based on whether the switch is on or off.
   * @param state The state of the AJSwitch child component.
   * @return {*} The classes attached to the 'on' icon.
   */
  const getOnIconClasses = (state) => {
    return state ?
      classnames(classes.ajSwitchIcon, 'aj-switch-icon-on') :
      classnames(classes.ajSwitchIcon, classes.ajSwitchIconNotSelected, 'aj-switch-icon-on');
  };

  // Hooks for the classes attached to the 'figure' elements.
  const [offIconClasses, setOffIconClasses] = useState(getOffIconClasses(initialState));
  const [onIconClasses, setOnIconClasses] = useState(getOnIconClasses(initialState));

  /**
   * Function called when the state of the AJSwitch child component changes.
   * @param state Whether the AJSwitch child component is turned on or off.
   */
  const onSwitchChange = (state) => {
    setOffIconClasses(getOffIconClasses(state));
    setOnIconClasses(getOnIconClasses(state));

    if (typeof onChange === 'function') {
      onChange(state);
    }
  };

  return (
    <div className={classnames(classes.ajSwitchIconContainer, 'aj-switch-icon')}>
      <figure className={offIconClasses}>
        <img src={offImageUrl} alt="" />
      </figure>
      <AJSwitch onChange={onSwitchChange} initialState={initialState} disabled={disabled} />
      <figure className={onIconClasses}>
        <img src={onImageUrl} alt="" />
      </figure>
    </div>
  );
};

AJSwitchIcon.propTypes = {
  offImageUrl: PropTypes.string.isRequired,
  onImageUrl: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  initialState: PropTypes.bool,
  disabled: PropTypes.bool
};

export default AJSwitchIcon;
