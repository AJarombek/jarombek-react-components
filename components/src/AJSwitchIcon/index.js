import AJSwitchIcon from './AJSwitchIcon';

export default AJSwitchIcon;
