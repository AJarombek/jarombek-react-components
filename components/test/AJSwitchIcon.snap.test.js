/**
 * Snapshot test for the {@link ../../components/AJSwitchIcon} component.
 * @author Andrew Jarombek
 * @since 1/1/2020
 */

import React from 'react';
import AJSwitchIcon from '../src/AJSwitchIcon/AJSwitchIcon';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <AJSwitchIcon
      offImageUrl="https://asset.jarombek.com/jarombek.png"
      onImageUrl="https://asset.jarombek.com/jarombek.png"
      onChange={f=>f}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with all props', () => {
  const tree = renderer.create(
    <AJSwitchIcon
      offImageUrl="https://asset.jarombek.com/jarombek.png"
      onImageUrl="https://asset.jarombek.com/jarombek.png"
      onChange={f=>f}
      initialState={true}
      disabled={true}/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
