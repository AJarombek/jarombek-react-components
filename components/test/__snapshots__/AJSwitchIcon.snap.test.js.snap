// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ajSwitchIconContainer-0-2-1 aj-switch-icon"
>
  <figure
    className="ajSwitchIcon-0-2-2 aj-switch-icon-off"
  >
    <img
      src="https://asset.jarombek.com/jarombek.png"
    />
  </figure>
  <div
    className="ajSwitch-0-2-4 ajSwitchInactive-0-2-7 aj-switch"
    onClick={[Function]}
  >
    <div
      className="ajSwitchHead-0-2-8 ajSwitchHeadInactive-0-2-10 aj-switch-head"
    >
       
    </div>
    <div
      className="ajSwitchTail-0-2-11 ajSwitchTailInactive-0-2-13 aj-switch-tail"
    >
       
    </div>
  </div>
  <figure
    className="ajSwitchIcon-0-2-2 ajSwitchIconNotSelected-0-2-3 aj-switch-icon-on"
  >
    <img
      src="https://asset.jarombek.com/jarombek.png"
    />
  </figure>
</div>
`;

exports[`renders correctly with all props 1`] = `
<div
  className="ajSwitchIconContainer-0-2-1 aj-switch-icon"
>
  <figure
    className="ajSwitchIcon-0-2-2 ajSwitchIconNotSelected-0-2-3 aj-switch-icon-off"
  >
    <img
      src="https://asset.jarombek.com/jarombek.png"
    />
  </figure>
  <div
    className="ajSwitch-0-2-4 ajSwitchActive-0-2-6 ajSwitchDisabled-0-2-5 aj-switch"
    onClick={[Function]}
  >
    <div
      className="ajSwitchHead-0-2-8 ajSwitchHeadActive-0-2-9 aj-switch-head"
    >
       
    </div>
    <div
      className="ajSwitchTail-0-2-11 ajSwitchTailActive-0-2-12 aj-switch-tail"
    >
       
    </div>
  </div>
  <figure
    className="ajSwitchIcon-0-2-2 aj-switch-icon-on"
  >
    <img
      src="https://asset.jarombek.com/jarombek.png"
    />
  </figure>
</div>
`;
